docs/csharp/whats-new/csharp-8.md

```diff
@@ -0,0 +1,314 @@
+
+You can learn more about the feature in the overview of [nullable reference types](../nullable-references.md). Try it yourself in a new application in this [nullable reference types tutorial](../tutorials/nullable-reference-types.md). Learn about the steps to migrate an existing codebase to make use of nullable reference types in the [migrating an application to use nullable reference types tutorial]../tutorials/upgrade-to-nullable-references.md).
+
+## Asynchronous streams
+
+Starting with C# 8.0, you can create and consume streams asynchronously. A method that returns an asynchronous stream has three properties:
+
+1. It was declared with the `async` modifier.
+1. It returns an <xref:System.Collections.Generic,IAsyncEnumerable%601>.
+1. The method contains `yield return` statements to return successive elements in the asynchronous stream.
+
+Consuming an asynchronous stream requires you to add the `await` keyword before the `foreach` keyword when you enumerate the elements of the stream. Adding the `await` keyword requires the method that enumerates the asynchronous stream to be declared with the `async` modifier and to return a type allowed for an `async` method. Typically that means returning a <xref:System.Threading.Tasks.Task> or <xref:System.Threading.Tasks.Task%601>. It can also be a <xref:System.Threading.Tasks.ValueTask> or <xref:System.Threading.Tasks.ValueTask%601>. A method can both consume and produce an asynchronous stream, which means it would return an <xref:System.Collections.Generic,IAsyncEnumerable%601>.
+
+You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md).
+
+## Ranges and indices
+
+Ranges and indices provide a succinct syntax for specifying subranges in an array, <xref:System.Span%601>, or <xref:System.ReadOnlySpan%601>. Consider the following array:
+
+```csharp
+var words = new string[]
+{
+    "The",
+    "quick",
+    "brown",
+    "fox",
+    "jumped",
+    "over",
+    "the",
+    "lazy",
+    "dog"
+};
+```
+
+You can retrieve the last word by prefixing the index `1` with `^`:
+
+```csharp
+Console.WriteLine($"The last word is {words[^1]}");
+```
+
+When you take only a single element and specify from the end, indexing starts at 1.
+
+You can also specify ranges of elements. The following code creates a subrange with the words "quick", "brown", and "fox":
+
+```csharp
+var brownFox = words[1..4];
+```
+
+Ranges can be specified from the end as well as the beginning. The following code creates a subrange with "lazy" and "dog":
+
+```csharp
+var lazyDog = words[^2..^0];
+```
+
+You can also create ranges that are open ended for the start, end, or both, as in the following code:
```

Comment:
It may be easier to start with this stuff because this is a little more familiar to those used to 0-based indexing. You may want to start with using the `Range` type directly and then say that there is some sort of implicit conversion going on with the range syntax? (if that is true, I just assume it is)